crypto/s2n_evp_signing.c

```diff
@@ -135,7 +137,7 @@ int s2n_evp_verify(const struct s2n_pkey *pub, s2n_signature_algorithm sig_alg,
     POSIX_ENSURE(s2n_evp_signing_supported(), S2N_ERR_HASH_NOT_READY);
     POSIX_GUARD_RESULT(s2n_evp_signing_validate_hash_alg(sig_alg, hash_state->alg));
 
-    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(pub->pkey, NULL);
+    DEFER_CLEANUP(EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(pub->pkey, NULL), EVP_PKEY_CTX_free_pointer);
```

Comment:
@torben-hansen brought up concerns about leaving the evp_md_ctx with a pointer to a freed evp_pkey_ctx. I /think/ we're currently technically fine because our hash states are consumed by the signing operation and not reused without being re-initalized. However, I could be missing something, we could change how we handle hash states later, and it's bad hygiene anyway.

From the [docs](https://www.openssl.org/docs/man1.1.1/man3/EVP_MD_CTX_set_pkey_ctx.html):
> The pctx passed to this function should be freed by the caller. A NULL pctx pointer is also allowed to clear the EVP_PKEY_CTX assigned to ctx. In such case, freeing the cleared EVP_PKEY_CTX or not depends on how the EVP_PKEY_CTX is created.

The safest thing to do is to call s2n_evp_md_ctx_set_pkey_ctx with a NULL before we actually free the pkey context. You could probably do it reliably by also wrapping `ctx` in a DEFER_CLEANUP that does the set.